cmake.md

```diff
@@ -0,0 +1,63 @@
+## CMake build system
+### Terminology
+
+The CMake system in symbiflow-arch-defs uses the following hierarchy:
+
+* `ARCH` - Architecture is the high level grouping of a set of "similar" FPGA
+  types. An example would be the "Lattice iCE40 Series" (ice40) or the "Xilinx Artix 7 Series" (artix7).
+* `DEVICE_TYPE` - A type of device.  This is defined as an `ARCH` and the
+  arch.xml which defines the units with this device type.
+* `DEVICE` - A specific instance of a `DEVICE_TYPE`.  A device definition is
+   a `DEVICE_TYPE`, a `PACKAGE` list, and IO map definitions for each package
+   if needed.  Each `DEVICE` should coorispond with a layout definition within
+   the `DEVICE_TYPE`.
+* `BOARD` - A specific `DEVICE` and `PACKAGE` instance, along with a
+  `PROG_TOOL`/`PROG_CMD` (e.g. commands to program a specified board).
+
+### Creating a new FPGA target
+
+Assuming that the board(s) you are interested in targetting is already
+defined, `ADD_FPGA_TARGET` is what will create targets.  See
+[`ADD_FPGA_TARGET` documentation](make/devices.cmake#L559).  `ADD_FPGA_TARGET`
+is roughly equivilant to CMake's `ADD_EXECUTABLE`.  Unlink `ADD_EXECUTABLE`,
+`ADD_FPGA_TARGET` requires that you specify which `BOARD` you are targetting.
+
+If you want to target multiple boards, `ADD_FPGA_TARGET_BOARDS` exists
+that will call `ADD_FPGA_TARGET` correctly for each board you request.  See
+[`ADD_FPGA_TARGET_BOARDS` documentation](make/devices.cmake#L458).
+
+#### Note on `ADD_FILE_TARGET` and `ADD_FPGA_BOARD`
+
+All source files in the symbiflow-arch-defs are required to have a file target
+associated with them.  This is done via
+[`ADD_FILE_TARGET`](make/file_targets.cmake#L193).  By default
+`ADD_FPGA_TARGET` and `ADD_FPGA_TARGET_BOARDS` will both implicitly invoke
+`ADD_FILE_TARGET` for you.  This handles the common case where input sources
+are not generated files.  However, if you want to generate files, then you
+cannot use the implicity `ADD_FILE_TARGET`, and you will need to add
+`ADD_FILE_TARGET` calls for all inputs to `ADD_FPGA_TARGET`.
+
+For non-generated verilog, `ADD_FILE_TARGET` should be invoked like:
+```
+add_file_target(FILE <current source dir relative path> SCANNER_TYPE verilog)
+```
+
+For generated verilog, `ADD_FILE_TARGET` should be invoked like:
+```
+add_file_target(FILE <current source dir relative path> GENERATED)
+```
+Note that the input path must be current source dir relative, not absolute,
```

Comment:
```
The input path must be relative to current source directory (the normal options of absolute and relative to cmake build directory will not work).

As CMake cannot take dependencies during build time, generated files thus must have their dependencies manually supplied via using the [`APPEND_FILE_DEPENDENCY`](make/file_targets.cmake#L79) call.

Because we have generated Verilog files and they use relative include paths, source files need to be copied into the build output directory. This mandates the usage of the `ADD_FILE_TARGET`.
```

BTW As mentioned on chat -- Yosys and iverilog both have ways to set the include paths... I'm unsure if that makes a difference here?